File: backend/src/routes/uploads.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const fs = require('fs').promises;
const { authenticateToken, requireRole } = require('../middleware/auth');
const router = express.Router();

// endpoint для скачивания файлов чата
router.get('/chats/:filename', 
  authenticateToken,
  requireRole(['SUPERADMIN', 'EX_ADMIN', 'MANAGER', 'OPERATOR']),
  async (req, res) => {
    try {
      const { filename } = req.params;
      const baseFilename = path.basename(String(filename || ''));
      const filePath = path.join(__dirname, '../../uploads/chats', baseFilename);
      
      // проверяем существование файла
      try {
        await fs.access(filePath);
      } catch {
        return res.status(404).json({
          success: false,
          message: 'Файл не найден'
        });
      }
      
      // определяем тип файла и отправляем
      const ext = path.extname(filename).toLowerCase();
      let contentType = 'application/octet-stream';
      
      if (['.jpg', '.jpeg'].includes(ext)) {
        contentType = 'image/jpeg';
      } else if (ext === '.png') {
        contentType = 'image/png';
      } else if (ext === '.pdf') {
        contentType = 'application/pdf';
      }
      
      res.setHeader('Content-Type', contentType);

      const safeFilename = baseFilename
        .replace(/[\r\n]/g, '')
        .replace(/["\\]/g, '_')
        .replace(/[^\x20-\x7E]+/g, '_')
        .trim() || 'file';
      const encodedFilename = encodeURIComponent(baseFilename || 'file');

      res.setHeader(
        'Content-Disposition',
        `inline; filename="${safeFilename}"; filename*=UTF-8''${encodedFilename}`
      );
      res.sendFile(filePath);
      
    } catch (error) {
      console.error('Ошибка скачивания файла:', error);
      res.status(500).json({
        success: false,
        message: 'Ошибка скачивания файла',
        error: error.message
      });
    }
  }
);

// настройка хранилища файлов
const storage = multer.diskStorage({
  destination: async (req, file, cb) => {
    const uploadDir = path.join(__dirname, '../../uploads/mailings');
    
    // создаем папку если не существует
    try {
      await fs.access(uploadDir);
    } catch {
      await fs.mkdir(uploadDir, { recursive: true });
    }
    
    cb(null, uploadDir);
  },
  filename: (req, file, cb) => {
    // генерируем уникальное имя файла
    const uniqueSuffix = Date.now() + '-' + Math.round(Math.random() * 1E9);
    const originalName = Buffer.from(file.originalname, 'latin1').toString('utf8');
    const ext = path.extname(originalName);
    const name = path.basename(originalName, ext);
    
    cb(null, `${name}-${uniqueSuffix}${ext}`);
  }
});

// фильтр типов файлов
const fileFilter = (req, file, cb) => {
  const allowedTypes = [
    'image/jpeg', 
    'image/jpg', 
    'image/png', 
    'image/gif', 
    'image/webp'
  ];
  
  if (allowedTypes.includes(file.mimetype)) {
    cb(null, true);
  } else {
    cb(new Error('Неподдерживаемый тип файла. Разрешены: JPEG, PNG, GIF, WebP'), false);
  }
};

// настройка multer
const upload = multer({
  storage: storage,
  limits: {
    fileSize: 10 * 1024 * 1024, // 10MB максимум
    files: 10 // максимум 10 файлов
  },
  fileFilter: fileFilter
});

// загрузка файлов для рассылки
router.post('/mailing-attachments', 
  authenticateToken,
  requireRole(['SUPERADMIN', 'EX_ADMIN']),
  upload.array('files', 10), // максимум 10 файлов
  async (req, res) => {
    try {
      if (!req.files || req.files.length === 0) {
        return res.status(400).json({
          success: false,
          message: 'Файлы не были загружены'
        });
      }

      // формируем ответ с информацией о загруженных файлах
      const uploadedFiles = req.files.map(file => ({
        filename: file.filename,
        originalName: file.originalname,
        mimetype: file.mimetype,
        size: file.size,
        path: `/uploads/mailings/${file.filename}`,
        url: `${req.protocol}://${req.get('host')}/uploads/mailings/${file.filename}`
      }));

      res.json({
        success: true,
        message: `Успешно загружено ${uploadedFiles.length} файл(ов)`,
        files: uploadedFiles
      });

    } catch (error) {
      console.error('Ошибка загрузки файлов:', error);
      res.status(500).json({
        success: false,
        message: 'Ошибка загрузки файлов',
        error: error.message
      });
    }
  }
);

// удаление файла
router.delete('/mailing-attachments/:filename',
  authenticateToken,
  requireRole(['SUPERADMIN', 'EX_ADMIN']),
  async (req, res) => {
    try {
      const { filename } = req.params;
      const filePath = path.join(__dirname, '../../uploads/mailings', filename);
      
      // проверяем существование файла
      try {
        await fs.access(filePath);
      } catch {
        return res.status(404).json({
          success: false,
          message: 'Файл не найден'
        });
      }
      
      // удаляем файл
      await fs.unlink(filePath);
      
      res.json({
        success: true,
        message: 'Файл успешно удален'
      });
      
    } catch (error) {
      console.error('Ошибка удаления файла:', error);
      res.status(500).json({
        success: false,
        message: 'Ошибка удаления файла',
        error: error.message
      });
    }
  }
);

// получение списка файлов (для отладки)
router.get('/mailing-attachments',
  authenticateToken,
  requireRole(['SUPERADMIN']),
  async (req, res) => {
    try {
      const uploadDir = path.join(__dirname, '../../uploads/mailings');
      
      try {
        const files = await fs.readdir(uploadDir);
        const filesInfo = [];
        
        for (const file of files) {
          const filePath = path.join(uploadDir, file);
          const stats = await fs.stat(filePath);
          
          filesInfo.push({
            filename: file,
            size: stats.size,
            created: stats.birthtime,
            modified: stats.mtime,
            url: `${req.protocol}://${req.get('host')}/uploads/mailings/${file}`
          });
        }
        
        res.json({
          success: true,
          files: filesInfo
        });
        
      } catch {
        res.json({
          success: true,
          files: []
        });
      }
      
    } catch (error) {
      console.error('Ошибка получения списка файлов:', error);
      res.status(500).json({
        success: false,
        message: 'Ошибка получения списка файлов',
        error: error.message
      });
    }
  }
);

module.exports = router;

File: python_backend/app/routers/uploads.py
import os
import uuid
import shutil
from pathlib import Path
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import require_roles
from app.models.support import Support

router = APIRouter(prefix="/api/uploads", tags=["uploads"])

BASE_UPLOAD_DIR = Path(__file__).parent.parent.parent / "uploads"
CHAT_UPLOAD_DIR = BASE_UPLOAD_DIR / "chats"
MAILING_UPLOAD_DIR = BASE_UPLOAD_DIR / "mailings"

ALLOWED_IMAGE_TYPES = {"image/jpeg", "image/jpg", "image/png", "image/gif", "image/webp"}


def _safe_filename(filename: str) -> str:
    return Path(filename).name


# ── GET /chats/{filename} ─────────────────────────────────────────────────────
@router.get("/chats/{filename}")
async def download_chat_file(
    filename: str,
    current_user: Support = Depends(require_roles("SUPERADMIN", "EX_ADMIN", "MANAGER", "OPERATOR")),
):
    safe = _safe_filename(filename)
    file_path = CHAT_UPLOAD_DIR / safe
    if not file_path.exists():
        raise HTTPException(404, "File not found")

    ext = file_path.suffix.lower()
    media_types = {
        ".jpg": "image/jpeg", ".jpeg": "image/jpeg",
        ".png": "image/png", ".pdf": "application/pdf",
    }
    media_type = media_types.get(ext, "application/octet-stream")
    return FileResponse(str(file_path), media_type=media_type,
                        headers={"Content-Disposition": f'inline; filename="{safe}"'})


# ── POST /mailing-attachments ─────────────────────────────────────────────────
@router.post("/mailing-attachments")
async def upload_mailing_attachments(
    files: list[UploadFile] = File(...),
    current_user: Support = Depends(require_roles("SUPERADMIN", "EX_ADMIN")),
):
    MAILING_UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

    uploaded = []
    for file in files:
        if file.content_type not in ALLOWED_IMAGE_TYPES:
            raise HTTPException(400, f"Unsupported file type: {file.content_type}. Allowed: JPEG, PNG, GIF, WebP")
        if file.size and file.size > 10 * 1024 * 1024:
            raise HTTPException(400, "File too large. Max 10MB")

        ext = Path(file.filename or "").suffix
        filename = f"{uuid.uuid4().hex}{ext}"
        dest = MAILING_UPLOAD_DIR / filename
        with open(dest, "wb") as f:
            shutil.copyfileobj(file.file, f)

        uploaded.append({
            "filename": filename,
            "originalName": file.filename,
            "mimetype": file.content_type,
            "size": dest.stat().st_size,
            "path": f"/uploads/mailings/{filename}",
        })

    return {"success": True, "message": f"Uploaded {len(uploaded)} file(s)", "files": uploaded}


# ── DELETE /mailing-attachments/{filename} ────────────────────────────────────
@router.delete("/mailing-attachments/{filename}")
async def delete_mailing_attachment(
    filename: str,
    current_user: Support = Depends(require_roles("SUPERADMIN", "EX_ADMIN")),
):
    safe = _safe_filename(filename)
    file_path = MAILING_UPLOAD_DIR / safe
    if not file_path.exists():
        raise HTTPException(404, "File not found")
    file_path.unlink()
    return {"success": True, "message": "File deleted"}


# ── GET /mailing-attachments ─────────────────────────────────────────────────
@router.get("/mailing-attachments")
async def list_mailing_attachments(
    current_user: Support = Depends(require_roles("SUPERADMIN")),
):
    MAILING_UPLOAD_DIR.mkdir(parents=True, exist_ok=True)
    files = []
    for p in MAILING_UPLOAD_DIR.iterdir():
        if p.is_file():
            stat = p.stat()
            files.append({
                "filename": p.name,
                "size": stat.st_size,
                "created": stat.st_ctime,
                "modified": stat.st_mtime,
                "path": f"/uploads/mailings/{p.name}",
            })
    return {"success": True, "files": files}
